cli/src/breakpoint_utils.py

```diff
@@ -0,0 +1,273 @@
+# will render it as an array."
+#
+# As the breakpoint IDs will used as keys in the RTDB we do not want the above
+# behaviour, as we always want maps to be returned instead of arrays.
+def get_new_breakpoint_id(user_output, firebase_rtdb_service, debuggee_id):
+  time_secs = int(time.time())
+  breakpoint_id = None
+  found = False
+
+  for i in range(0, 10):
+    breakpoint_id = f'b-{time_secs + i}'
+    active_path = f'breakpoints/{debuggee_id}/active/{breakpoint_id}'
+    final_path = f'breakpoints/{debuggee_id}/final/{breakpoint_id}'
+
+    bp_active = firebase_rtdb_service.get(active_path, shallow=True)
+    bp_final = firebase_rtdb_service.get(final_path, shallow=True)
+
+    # This case means there are no breakpoints at all.
+    if bp_active is None and bp_final is None:
+      found = True
+      break
+
+  if not found:
+    user_output.error(
+        'ERROR Failed to determine a new breakpoint ID, please try again')
+    raise SilentlyExitError
+
+  return breakpoint_id
+
+
+def convert_unix_msec_to_rfc3339(unix_msec):
+  """Converts a Unix timestamp represented in milliseconds since the epoch to an
+
+  RFC3339 string representation.
+
+  Args:
+    unix_msec: The Unix timestamp, represented in milliseconds since the epoch.
+
+  Returns:
+    An RFC3339 encoded timestamp string in format: "%Y-%m-%dT%H:%M:%S.%fZ".
+  """
+  seconds = unix_msec / 1000
+  msec = unix_msec % 1000
+  timestamp = seconds + msec / float(1000)
+  dt = datetime.datetime.utcfromtimestamp(timestamp)
+  return dt.strftime('%Y-%m-%dT%H:%M:%S.%f') + 'Z'
+
+
+def set_converted_timestamps(bp):
+  conversions = [['createTime', 'createTimeUnixMsec'],
+                 ['finalTime', 'finalTimeUnixMsec']]
+
+  for c in conversions:
+    if c[0] not in bp and c[1] in bp:
+      bp[c[0]] = convert_unix_msec_to_rfc3339(bp[c[1]])
+
+  return bp
+
+
+# Returns None if there's an issue
+def normalize_breakpoint(bp, bpid=None):
+  if bp is None:
+    return None
+
+  if 'id' not in bp and bpid is not None:
+    bp['id'] = bpid
+
+  required_fields = ['id', 'location']
+  for f in required_fields:
+    if f not in bp:
+      return None
+
+  required_location_fields = ['path', 'line']
+  for f in required_location_fields:
+    if f not in bp['location']:
+      return None
+
+  if 'action' not in bp:
+    bp['action'] = 'CAPTURE'
+
+  if 'isFinalState' not in bp:
+    bp['isFinalState'] = False
+
+  if 'createTimeUnixMsec' not in bp:
+    # Assuming everything is working correctly the createTimeUnixMsec value
+    # should be present, if not initialize it to 0 so it's set to something,
+    # but it's clear the time was not actually known.
+    bp['createTimeUnixMsec'] = 0
+
+  if 'finalTimeUnixMsec' not in bp and bp['isFinalState']:
+    bp['finalTimeUnixMsec'] = 0
+
+  if 'userEmail' not in bp:
+    bp['userEmail'] = 'unknown'
+
+  set_converted_timestamps(bp)
+
+  return bp
+
+
+def get_breakpoints_by_state(firebase_rtdb_service, bp_state, debuggee_id,
```

Comment:
It feels a bit odd to see these rtdb crud functions in a util file that's otherwise more traditional util functions.  Maybe it makes sense to move those crud functions into the firebase service or into a wrapper around the service.

Reply:
Good callout, I created https://github.com/GoogleCloudPlatform/snapshot-debugger/issues/9 to address this.